### PR TITLE
MGMT-3439 - Add S.M.A.R.T. information to storage triage signature

### PR DESCRIPTION
**Pull request by omertuc**
@oshercc @ronniel1

### CODE REVIEW

File: tools/add_triage_signature.py [hunk 1]
```diff
@@ -383,8 +384,87 @@ class StorageDetailSignature(Signature):
     def __init__(self, jira_client):
         super().__init__(jira_client, comment_identifying_string="h1. Host storage details:")
 
-    def _update_ticket(self, url, issue_key, should_update=False):
+    class SmartctlExitCode(Flag):
+        """
+        Taken from `man smartctl` "EXIT STATUS" section
+        """
```
Comment (osherdp) [inline tools/add_triage_signature.py:390]: Single line

File: tools/add_triage_signature.py [hunk 2]
```diff
@@ -383,8 +384,87 @@ class StorageDetailSignature(Signature):
...
+        exit_code = smartctl["exit_status"]
+
+        if exit_code != 0:
+            output.append("non-zero smartctl exit code: " + ", ".join(flag.name for flag in cls.SmartctlExitCode
+                                                                      if flag in cls.SmartctlExitCode(exit_code)))
+            if cls.SmartctlExitCode(exit_code) == cls.SmartctlExitCode.SOME_ATA_COMMAND_FAILED_OR_SMART_CHECKSUM_ERROR:
+                if "smart_status" not in smart:
+                    output.append(
+                        "{color:green}*this typically happens on virtual disks*{color} that don't actually have S.M.A.R.T. information,"
+                        " but may also happen due to other reasons")
```
Comment (osherdp) [inline tools/add_triage_signature.py:413]: So I wonder if green is the right color in this case
Maybe no color is the right thing here

File: tools/add_triage_signature.py [hunk 3]
```diff
@@ -383,8 +384,87 @@ class StorageDetailSignature(Signature):
...
+
+        if "messages" in smartctl:
+            for message in smartctl["messages"]:
```
Comment (osherdp) [inline tools/add_triage_signature.py:416]:
```
for message in smartctl.get("messages", []):
```

File: tools/add_triage_signature.py (hunk 2 above)
Comment (osherdp) [reply tools/add_triage_signature.py:413]: or some other neutral color

File: tools/add_triage_signature.py [hunk 4]
```diff
@@ -383,8 +384,87 @@ class StorageDetailSignature(Signature):
...
+
+        if "power_on_time" in smart:
+            if "hours" in smart["power_on_time"]:
+                hours = smart["power_on_time"]["hours"]
+                output.append(
+                    f"disk has been powered on for {hours // 24} days and {hours % 24} hours throughout its life")
+
+        if "ata_smart_attributes" in smart:
+            table = smart["ata_smart_attributes"].get("table", [])
+            interesting = ("Program_Fail_Count", "Erase_Fail_Count", "Offline_Uncorrectable")
```
Comment (osherdp) [inline tools/add_triage_signature.py:443]: maybe rename to ``relevant_attributes``

File: tools/add_triage_signature.py [hunk 5]
```diff
@@ -383,8 +384,87 @@ class StorageDetailSignature(Signature):
...
+        return ", ".join(output)
+
+    @classmethod
+    def _parse_smart(cls, smart_json):
+        try:
+            return cls._parse_smart_internal(json.loads(smart_json))
+        except json.JSONDecodeError as e:
+            return f"Failed to decode S.M.A.R.T. JSON, error: {e}"
+        except Exception as e:
+            return f"S.M.A.R.T. JSON has unexpected format, error: {e}"
```
Comment (osherdp) [inline tools/add_triage_signature.py:465]: above exception should be JSON decode error, no?

File: tools/add_triage_signature.py [hunk 6]
```diff
@@ -396,19 +476,24 @@ def _update_ticket(self, url, issue_key, should_update=False):
...
             hosts.append(OrderedDict(
                 id=host['id'],
                 hostname=self._get_hostname(host),
                 disk_name="\n".join(disks_details['name']),
                 disk_type="\n".join(disks_details['type']),
                 disk_path="\n".join(disks_details['path']),
                 disk_bootable="\n".join(disks_details['bootable']),
-                disk_by_path="\n".join(disks_details['by-path'])
+                disk_by_path="\n".join(disks_details['by-path']),
+                SMART="\n".join(disks_details['smart']),
```
Comment (osherdp) [inline tools/add_triage_signature.py:496]: even though it's an acronym, I think it should stay ``smart`` because of Python conventions

**Review by osherdp — CHANGES_REQUESTED** [review]

File: tools/add_triage_signature.py (hunk 2 above)
Comment (omertuc) [reply tools/add_triage_signature.py:413]: It's a long sentence, and it generally means "Everything is probably OK, this is probably just a virtual disk, SMART is uninteresting". I want to draw the user to the "this is okay part" to reduce mental effort when going over this long table of disks.

File: tools/add_triage_signature.py (hunk 4 above)
Comment (omertuc) [reply tools/add_triage_signature.py:443]: Meet you halfway `interesting_attributes`

File: tools/add_triage_signature.py (hunk 5 above)
Comment (omertuc) [reply tools/add_triage_signature.py:465]: No, the second message I meant that the structure of the JSON is unexpected, not that the JSON encoding itself is wrong.

File: tools/add_triage_signature.py (hunk 6 above)
Comment (omertuc) [reply tools/add_triage_signature.py:496]: This kwarg is directly used as the header of the table shown to the user, and just lowercase "smart" is confusing - not everybody knows this acronym

File: tools/add_triage_signature.py (hunk 5 above)
Comment (omertuc) [reply tools/add_triage_signature.py]: Improved, somewhat

File: tools/add_triage_signature.py (hunk 6 above)
Comment (osherdp) [reply tools/add_triage_signature.py:493]: So change to ``"SMART": "value"``` (other values as well. Although it's weird we show ``disk_name`` and not ``Disk Name`` in a human readable table)

Comment (omertuc) [reply tools/add_triage_signature.py:493]: Not sure I understand

Comment (osherdp) [reply tools/add_triage_signature.py:493]:
```suggestion
                "Disk Path": "\n".join(disks_details['path']),
                ...
                "SMART": "\n".join(disks_details['smart']),
```

**Review by osherdp — APPROVED** [review]

**Pull request merged**